pennylane/allocation.py

```diff
@@ -0,0 +1,275 @@
+
+    ..see-also:: :func:`~.allocate`, :func:`~.allocate_ctx`.
+
+    """
+
+    def __init__(self, wires, require_zeros=True):
+        super().__init__(wires=wires)
+        self._hyperparameters = {
+            "require_zeros": require_zeros,
+        }
+
+    @property
+    def require_zeros(self):
+        """Whether or not the allocated wires are required to be in the zero state."""
+        return self.hyperparameters["require_zeros"]
+
+    @classmethod
+    def from_num_wires(cls, num_wires: int, require_zeros=True) -> "Allocate":
+        """Initialize an ``Allocate`` op from a number of wires instead of already constructed dynamic wires."""
+        wires = tuple(DynamicWire() for _ in range(num_wires))
+        return cls(wires=wires, require_zeros=require_zeros)
+
+
+class Deallocate(Operator):
+    """An instruction to deallocate the provided ``DynamicWire``'s.
+
+    Args:
+        wires (DynamicWire, Sequence[DynamicWire]): one or more dynamic wires to deallocate.
+        reset_to_original (bool): Whether or not the qubit will be in the same state upon being freed.
+
+    """
+
+    def __init__(self, wires: DynamicWire | Sequence[DynamicWire], reset_to_original=False):
+        super().__init__(wires=wires)
+        self._hyperparameters = {"reset_to_original": reset_to_original}
+
+    @property
+    def reset_to_original(self):
+        """Whether or not the dynamic wire will be returned to its original state."""
+        return self.hyperparameters["reset_to_original"]
+
+
+def allocate(num_wires: int, require_zeros: bool = True) -> Wires:
+    """Allocate dynamic wires for temporary use within a circuit.
+
+    .. warning::
+
+        This feature is experimental and may not be possible on all devices.
+
+
+    Args:
+        num_wires (int): the number of wires to allocate
+
+    Keyword Args:
+        require_zeros (bool): Whether or not the wires must start in the ``0`` state.
+
+    Returns:
+        Wires: A wires object containing ``DynamicWire`` objects.
+
+    .. see-also:: :class:`~.allocate_ctx`
+
+    :class:`~.allocate_ctx` is recommended as the preferred way to allocate wires, as it enforces automatic deallocation.
+    Manual use of ``allocate`` and ``deallocate`` should be used with more deliberate care.
+
+    ..code-block:: python
+
+        @qml.qnode(qml.device('default.qubit'))
+        def c():
+            qml.H(0)
+
+            wires = qml.allocation.(1, require_zeros=True)
+            qml.CNOT((0, wires[0]))
+            qml.CNOT((0, wires[0]))
+            qml.allocation.deallocate(wires, reset_to_original=True)
+
+            new_wires = qml.allocation.(1)
+            qml.SWAP((0, new_wires[0]))
+            qml.allocation.deallocate(new_wires)
+
+            return qml.probs(wires=0)
+
+        print(qml.draw(c, level="user")())
+
+
+    >>> print(qml.draw(c, level="user")())
+                0: ──H────────╭●─╭●─────────────╭SWAP─────────────┤  Probs
```

Comment:
I _think_ Sphinx complains about this indentation.

Reply:
```suggestion
    0: ──H────────╭●─╭●─────────────╭SWAP─────────────┤  Probs
```